messaging/src/main/java/io/lantern/messaging/Messaging.kt

```diff
@@ -246,33 +256,49 @@ class Messaging(
      *                       (application IDs are added to existing set)
      * @param minimumVerificationLevel the contact's verification will be set to the greater of this
      *                                 or the current verificationLevel (defaults to UNVERIFIED)
+     * @param chatNumber the ChatNumber to associate with this contact
+     *                   (if unspecified, the existing ChatNumber is left alone)
      * @param updateApplicationData optional function to manipulate application specific data on the
      *                              contact
      * @return the created or updated Contact
      */
     @Throws(
         InvalidKeyException::class,
         InvalidCharacterException::class,
+        java.lang.IllegalArgumentException::class,
     )
     fun addOrUpdateDirectContact(
-        unsafeId: String,
+        unsafeId: String? = null,
         displayName: String? = null,
         source: Model.ContactSource? = null,
         applicationIds: Map<Int, String>? = null,
         minimumVerificationLevel: Model.VerificationLevel =
             Model.VerificationLevel.UNACCEPTED,
+        chatNumber: Model.ChatNumber? = null,
         updateApplicationData: ((MutableMap<String, Any>) -> Unit)? = null
-    ): Model.Contact =
-        cryptoWorker.submitForValue {
+    ): Model.Contact {
+        if (unsafeId == null && chatNumber == null) {
+            throw java.lang.IllegalArgumentException("Please specify either unsafeId or chatNumber")
+        }
+
+        val contactId = when {
+            unsafeId != null -> unsafeId.directContactId
```

Comment:
is there a preference between `unsafeId` and `chatNumber` for how the client invokes this?

Reply:
The short answer is to use what you've got.

In practice, if you scanned a barcode that has the contact id, use that. If someone entered a full chat number, use that. If someone entered a short chat number, use the API look up the full ChatNumber and use that.

Reply:
I've added some logic to make this safer. If you provide an incomplete ChatNumber, it'll happily use it to derive the ContactId but won't record it on the contact. That way, we'll go fetch the real complete chat number from the server.